### PR TITLE
STITCH-1366 - Implement Twilio Service in Core, Android, and Server

### PR DESCRIPTION
**Pull request by edaniels**
In addition to Twilio...

- Clean up some tests
- Add some more tests
- Refactor packages to support publishing
- Allow direct construction of credentials to streamline auth flow
- Rename ClientProvider's to ClientFactory's
- Add mongodb-local to server
- Change Blocking suffix to Internal to reflect what it's really for
- Make builders more fluent (at the cost of duplication)

### CODE REVIEW

**Comment by edaniels** [comment]
@adamchel @jsflax please take a look when you get a chance and take your time. Happy to go over all of this in person too. It's a lot of changes but they are all pretty consistent and nothing should be too alarming.

File: core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAuthDocRequest.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    }
+
+    public Document getDocument() {
+      return this.document;
+    }
+
+    /**
+     * Set if this request should use an access token in this request.
+     */
+    public Builder withAccessToken() {
```
Comment (edaniels) [inline core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAuthDocRequest.java]: You'll see this pattern of overriding and using super (without the annotation) to indicate that this is purely for allowing you to do `builder.withSuperMethod().withConcreteMethod().withSuperMethod().build()` and still get back the concrete built type. CRTP (https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) didn't work out here and it confuses me very much in general :)

File: android/services/twilio/src/androidTest/java/com/mongodb/stitch/android/services/twilio/TwilioServiceClientIntTests.kt [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Assert.fail
+import org.junit.Assume.assumeTrue
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import java.util.concurrent.ExecutionException
+
+@RunWith(AndroidJUnit4::class)
```
Comment (edaniels) [inline android/services/twilio/src/androidTest/java/com/mongodb/stitch/android/services/twilio/TwilioServiceClientIntTests.kt]: @jsflax here are the int tests

File: core/services/twilio/src/test/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClientUnitTests.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+import static org.mockito.Mockito.verify;
+
+import com.mongodb.stitch.core.services.internal.CoreStitchService;
+import java.util.List;
+import org.bson.Document;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+public class CoreTwilioServiceClientUnitTests {
```
Comment (edaniels) [inline core/services/twilio/src/test/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClientUnitTests.java]: @jsflax and here are the unit tests.

**Comment by edaniels** [comment]
If we agree on changes here, we should make the changes on swift and js to maintain symmetry.

File: android/services/twilio/src/androidTest/java/com/mongodb/stitch/android/services/twilio/TwilioServiceClientIntTests.kt [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                app.second,
+                "twilio",
+                "twilio1",
+                ServiceConfigs.Twilio(getTwilioSid(), getTwilioAuthToken()))
+        addRule(svc.second, RuleCreator.Twilio("default", setOf(TwilioActions.Send)))
+
+        val client = getAppClient(app.first)
+        Tasks.await(client.auth.loginWithCredential(AnonymousCredential()))
+
+        val twilio = client.getServiceClient(TwilioServiceClient.Factory, "twilio1")
```
[the remark this answers is not in the document]
Comment (jsflax) [reply android/services/twilio/src/androidTest/java/com/mongodb/stitch/android/services/twilio/TwilioServiceClientIntTests.kt]: Should be fine on Swift side as well

File: core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/Response.java [hunk 5]
```diff
@@ -20,13 +20,15 @@
...
+import java.util.HashMap;
+import java.util.Locale;
 import java.util.Map;
 
 /**
  * A response to an HTTP request, typically originating from a {@link Request}.
  */
 public final class Response {
-  private final Integer statusCode;
+  private final int statusCode;
```
Comment (jsflax) [inline core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/Response.java]: Nice, always preferably to use primitives where possible, especially on mobile.

File: core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAuthRequest.java [hunk 6]
```diff
@@ -125,11 +127,67 @@ public Builder withShouldRefreshOnFailure(final boolean shouldRefresh) {
...
+    public Builder withBody(final byte[] body) {
+      super.withBody(body);
+      return this;
+    }
+
     /**
      * Builds the request.
      */
     public StitchAuthRequest build() {
+      if (useRefreshToken) {
```
Comment (jsflax) [inline core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchAuthRequest.java]: [nice]

File: core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchRequest.java [hunk 7]
```diff
@@ -190,4 +191,40 @@ public StitchRequest build() {
           method, path, headers == null ? new HashMap<String, String>() : headers, body, startedAt);
     }
   }
+
+  protected Document toDocument() {
```
Comment (jsflax) [inline core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchRequest.java]: [q] Why add this?

File: core/services/twilio/src/test/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClientUnitTests.java (hunk 3 above)
Comment (jsflax) [reply core/services/twilio/src/test/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClientUnitTests.java]: 👍

File: core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClient.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+public class CoreTwilioServiceClient {
+
+  private final CoreStitchService service;
+
+  protected CoreTwilioServiceClient(final CoreStitchService service) {
+    this.service = service;
+  }
+
+  private void sendMessageBlocking(
```
Comment (jsflax) [inline core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClient.java]: [nit][q] Why the blocking and internal?

File: core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/TwilioService.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.stitch.core.services.twilio;
+
+public final class TwilioService {
```
Comment (jsflax) [inline core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/TwilioService.java]: [q] Where is this used?

File: core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchRequest.java (hunk 7 above)
Comment (edaniels) [reply core/sdk/src/main/java/com/mongodb/stitch/core/internal/net/StitchRequest.java]: For tests. It makes for nice assertEquals :)

File: core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClient.java (hunk 8 above)
Comment (edaniels) [reply core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/CoreTwilioServiceClient.java]: whoops. It should be merged with the caller

File: core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/TwilioService.java (hunk 9 above)
Comment (edaniels) [reply core/services/twilio/src/main/java/com/mongodb/stitch/core/services/twilio/TwilioService.java]: I was thinking the admin SDK, but maybe this is better removed!

**Comment by edaniels** [comment]
Addressed comments. I think the remote MDB service within core will actually work really well. It will let us define everything in blocking behavior and have all common logic there. The android side will really just be one large facade that implements an android interface by extending core and wrapping its calls in dispatchers.

**Comment by edaniels** [comment]
@adamchel RFAL

File: android/services/twilio/src/main/java/com/mongodb/stitch/android/services/twilio/TwilioServiceClient.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+  /**
+   * Sends an SMS/MMS message.
+   *
+   * @param to The number to send the message to.
+   * @param from The number that the message is from.
+   * @param body The body text of the message.
+   * @return A task that completes when the send is done.
+   */
+  Task<Void> sendMessage(
```
[the remark this answers is not in the document]
Comment (edaniels) [reply android/services/twilio/src/main/java/com/mongodb/stitch/android/services/twilio/TwilioServiceClient.java]: I agree, let's add those kinds later.

**Pull request merged**